### DISCUSSION TITLE
Using Prisma in NextJS 13 AppDir

### DISCUSSION THREAD

**Discussion by TheLexoPlexx** [vercel/next.js#47876] [Help] [answered]
### Summary

Sort of related to https://github.com/vercel/next.js/discussions/40416

Alright, so what I came across is the following problem. I am using Prisma a lot in my current project and that of course includes server components to retrieve data but also client components to update data.

Especially using it together with forms is quite painful, not handling the form itself, but basically having to create an api-route for everything I am updating/creating in the db. Is there a more elegant solution to this? I am currently stuck with tons of boilerplate in my api-Directory.

The rich, amazing and especially useful type-system from Prisma is just lost and replaced with annoying "any".

Or maybe, I'm just doing it all wrong. See below:

### Additional information

```js
//pages/api/user/update.ts

type Data = {
  user?: User,
  message?: string
}

export default async function UpdateUser(
  req: NextApiRequest,
  res: NextApiResponse<Data>
) {
  return new Promise((resolve, reject) => {

    let updateData: UserMod = {
      ...req.body,
      birthday: new Date(req.body.birthday),
    }
    
    prismadb.user.update({
      where: {
        id: req.body.id
      },
      data: updateData
    }).then((user) => {
      console.log("User updated", user)
      res.status(200).json({ user: user })
      resolve(user)
      return
    }).catch((error) => {
      res.status(500).json({ message: error.message })
      reject(error.message)
      return
    })
  })
}
```

combined with a standard...
```tsx
    fetch("/api/user/update", {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify(formData)
    }).then(response => response.json())
    .then(data => {
      if (data.message) {
        alert(data.message)
        return
      } else {
        let user: User = data.user
        console.log(user)
        router.push("/team/" + user.id)
      }
    }).catch(error => {
      console.error(error)
    })
```

**Comment by theonlydaleking** [comment]
It's worth remembering that you are still sending data from client => Server. 

Magic happens now with 13 in terms of the Server => client flow but going the other way still needs to be an API call in order to follow all the security protocols. 

I know it's easy to want to be able to update prisma client side but that would involve exposing all of Prisma's keys to the client.

**Comment by TheLexoPlexx** [comment, marked as answer]
Next.js 13.4 solves this with Server Functions: https://nextjs.org/blog/next-13-4#server-actions-alpha

The way I did it was just a quick test and should NOT be copied straight away, things will probably change in the future. What I did, very rudimentary at least is the following:

```tsx
//page.tsx
"use client"

imports[..]

export default function NewTeam() {

  function handleChange(event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) {
    const { name, value } = event.target
    setFormData(prevState => ({ ...prevState, [name]: value }))
  }

  const [formData, setFormData] = useState({
    mail: "",
    vorname: "",
    name: "",
    password: randomPassword()
  })

  async function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault()

    serverCreateUser(formData)
  }

  return (
    <form onSubmit={handleSubmit}>
      <label>Mail:
        <input type="email" name="mail" id="mail" value={formData.mail} onChange={handleChange} />
      </label>

      <label>Surname:
        <input type="text" name="surname" id="surname" value={formData.vorname} onChange={handleChange} />
      </label>

      <label>Name:
        <input type="text" name="name" id="name" value={formData.name} onChange={handleChange} />
      </label>

      <label>Password:
        <input type="text" name="password" id="password" value={formData.password} onChange={handleChange} />
      </label>
      <button type="submit">Save</button>
    </form>
  )
}
```

And the library-function, directly using serverCreateUser:
```ts
//lib/createUser.ts
"use server"

import prisma from "lib/prismadb"

export async function serverCreateUser(formData: any) {
  return await prisma.user.create({
    data: {
      email: formData.mail,
      password: formData.password,
      name: formData.name,
      vorname: formData.vorname,
    }
  })
}
```

**Reply by theonlydaleking** [reply]
I stand corrected - I was not aware that React had been working on RPC's and that next was looking to add this as a Data Primitive. 

Probably still best to use the API pattern until this is stable. 

Thanks for the heads up on Server Mutations.
